src/serverenvironment.cpp

```diff
@@ -818,6 +838,15 @@ class ABMHandler
 		{
 			const MapNode &n = block->getNodeUnsafe(p0);
 			content_t c = n.getContent();
+			// cache content types as we go
+			if (!block->contents_cached && !block->do_not_cache_contents) {
+				block->contents.insert(c);
+				if (block->contents.size() > 64) {
```

Comment:
a thread local g_setting can be nice to use here

Reply:
I don't see a need for a setting.

Reply:
Why not? If you have lots of RAM to spare, this would be one of the many small adjustments you could make to improve the server's responsiveness.

Reply:
Nah. This is just a heuristic, as long as node types for "most" blocks can be cached this is good.
A block can have at most 16^3 = 4096 different node types. Caching at most 1/64th seems good.

1. If we cache too much there's additional memory *and* additional CPU to check the cache first
2. The more different node types we have the more likely it becomes that ABMs need to be run for this block. And this check is entirely in vain.

16, 32, or 64 are good. Let's not add another config option to MT, it has *way* too many already.